test_runner/performance/test_bulk_insert.py

```diff
@@ -28,3 +30,30 @@ def test_bulk_insert(neon_with_baseline: PgCompare):
 
             env.report_peak_memory_use()
             env.report_size()
+
+    # When testing neon, also check how long it takes the pageserver to reingest the
+    # wal from safekeepers. If this number is close to total runtime, then the pageserver
+    # is the bottleneck.
+    if isinstance(env, NeonCompare):
+        measure_recovery_time(env)
+
+
+def measure_recovery_time(env: NeonCompare):
+    client = env.env.pageserver.http_client()
+    pg_version = PgVersion(client.timeline_detail(env.tenant, env.timeline)["pg_version"])
+
+    # Stop pageserver and remove tenant data
+    env.env.pageserver.stop()
+    timeline_dir = env.env.timeline_dir(env.tenant, env.timeline)
+    shutil.rmtree(timeline_dir)
+
+    # Start pageserver
+    env.env.pageserver.start()
+
+    # Measure recovery time
+    with env.record_duration("wal_recovery"):
+        # Create the tenant, which will start walingest
+        client.timeline_create(pg_version, env.tenant, env.timeline)
+
+        # Flush, which will also wait for lsn to catch up
+        env.flush()
```

Comment:
Should we stop the endpoint after the first flush, so that it will produce no WAL more, and these two would operate near the same LSN to be comparable?

Could record those two lsns which we expect to be the same as well.

Aah but then we should had saved it on the first flush... I am afraid these abstractions might be too simple.

Reply:
Yeah I would add an lsn to flush(), but can't pass an lsn to the postgres implementation, `checkpoint`. Let me know if you see a way around it.